src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -185,121 +166,192 @@ def __enter__(self):
+                    if data == TERMINATE_SIGNAL or line_timeout_sec is None:
+                        bulk_logger.info(
+                            f"The thread monitoring the process [{process_id}-{process_name}] will be terminated."
+                        )
+                        # Put the terminate signal into the input queue to notify the sub process to exit.
+                        input_queue.put(TERMINATE_SIGNAL)
+                        # End the process if found the terminate signal.
+                        self._processes_manager.end_process(index)
+                        # In fork mode, the main process and the sub spawn process communicate through _process_info.
+                        # We need to ensure the process has been killed before returning. Otherwise, it may cause
+                        # the main process have exited but the spawn process is still alive.
+                        # At this time, a connection error will be reported.
+                        self._processes_manager.ensure_process_terminated_within_timeout(process_id)
+                        # Set exit_loop to True to exit the main loop.
+                        exit_loop = True
+                        break
+                    # If the task is a line execution request, put the request into the input queue.
+                    run_id, line_number, inputs = data
+                    args = (run_id, line_number, inputs, line_timeout_sec)
+                    input_queue.put(args)
                     break
-                line_timeout_sec = min(line_timeout_sec, remaining_execution_time)
-
-            # Put task into input_queue
-            args = (inputs, line_number, run_id, line_timeout_sec)
-            input_queue.put(args)
+                except queue.Empty:
+                    pass
 
-            self._processing_idx[line_number] = format_current_process_info(process_name, process_id, line_number)
-            log_process_status(process_name, process_id, line_number)
+            if exit_loop:
+                break
 
             start_time = datetime.utcnow()
             completed = False
             crashed = False
             returned_node_run_infos = {}
 
+            self._processing_idx[line_number] = format_current_process_info(process_name, process_id, line_number)
+            log_process_status(process_name, process_id, line_number)
             # Responsible for checking the output queue messages and processing them within a specified timeout period.
-            while not self._batch_timeout_expired(batch_start_time) and not self._line_timeout_expired(start_time):
+            while not self._line_timeout_expired(start_time, line_timeout_sec=line_timeout_sec):
                 # Monitor process aliveness.
-                crashed = not self._is_process_alive(process_id)
+                crashed = not self._processes_manager.is_process_alive(process_id)
                 if crashed:
                     break
```

Comment:
Overall we may try to make the whole function more clean, to show the core logic, maybe something like this:
```
while not terminated:
    task = get_task()
    if terminated:
        terminate_process()
    put_task_to_input_queue()
    output = get_result_from_output_queue(timeout=xxx)
    handle_output(output)
```

Reply:
Currently the code in this function is not that clear to show the main loop.

Reply:
Will refine in teh next PR.